src/http_interface.py

```diff
@@ -0,0 +1,162 @@
+
+import json
+import logging
+
+from ops import RelationBrokenEvent, RelationChangedEvent, RelationJoinedEvent
+from ops.charm import CharmBase, CharmEvents, RelationEvent
+from ops.framework import EventSource, Object
+from ops.model import ModelError, Relation, RelationDataContent
+
+import legacy
+
+logger = logging.getLogger()
+SERVICES_CONFIGURATION_KEY = "services"
+
+
+class HTTPDataProvidedEvent(RelationEvent):
+    """Event representing that http data has been provided."""
+
+
+class HTTPDataRemovedEvent(RelationEvent):
+    """Event representing that http data has been removed."""
+
+
+class HTTPProviderEvents(CharmEvents):
+    """Container for HTTP Provider events.
+
+    Attrs:
+        data_provided: Custom event when integration data is provided.
+        data_removed: Custom event when integration data is removed.
+    """
+
+    data_provided = EventSource(HTTPDataProvidedEvent)
+    data_removed = EventSource(HTTPDataRemovedEvent)
+
+
+class _IntegrationInterfaceBaseClass(Object):
+    """Base class for integration interface classes.
+
+    Attrs:
+        relations: The list of Relation instances associated with the charm.
+    """
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Initialize the interface base class.
+
+        Args:
+            charm: The charm implementing the requirer or provider.
+            relation_name: Name of the integration using the interface.
+        """
+        super().__init__(charm, relation_name)
+
+        observe = self.framework.observe
+        self.charm: CharmBase = charm
+        self.relation_name = relation_name
+
+        observe(charm.on[relation_name].relation_created, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+        observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_departed, self._on_relation_changed)
+        observe(charm.on[relation_name].relation_broken, self._on_relation_broken)
+
+    def _on_relation_joined(self, _: RelationJoinedEvent) -> None:
+        """Abstract method to handle relation-joined event."""
+
+    def _on_relation_changed(self, _: RelationChangedEvent) -> None:
+        """Abstract method to handle relation-changed event."""
+
+    def _on_relation_broken(self, _: RelationBrokenEvent) -> None:
+        """Abstract method to handle relation-changed event."""
+
+    @property
+    def relations(self) -> list[Relation]:
+        """The list of Relation instances associated with the charm."""
+        return list(self.charm.model.relations[self.relation_name])
+
+
+class HTTPProvider(_IntegrationInterfaceBaseClass):
+    """HTTP interface provider class to be instantiated by the haproxy-operator charm.
+
+    Attrs:
+        on: Custom events that are used to notify the charm using the provider.
+    """
+
+    on = HTTPProviderEvents()  # type: ignore
```

Comment:
All `type: ignore` and other linting disable rules require an explanatory comment please

Reply:
This is so that future readers know the reason for the disable.